### PR TITLE
CMS API:  Drop support for XML requests

### PR DESCRIPTION
**Pull request by thalesmiguel**
**What this PR does / why we need it**:

This PR removes support for `XML` format requests from all CMS endpoints. It also makes `JSON` the standard request format for the CMS API. 

**Which issue(s) this PR fixes** 

[THREESCALE-9433](https://issues.redhat.com/browse/THREESCALE-9433)

**Verification steps** 

The following endpoints should only respond to `JSON` requests. Any other format should return a `406` status.

| Verb | Endpoint | Controller#Action |
| ------- | ------- | ------- |
|GET|/admin/api/cms/sections/:section_id/files|admin/api/cms/files#index|
|GET|/admin/api/cms/sections|admin/api/cms/sections#index|
|POST|/admin/api/cms/sections|admin/api/cms/sections#create|
|GET|/admin/api/cms/sections/new|admin/api/cms/sections#new|
|GET|/admin/api/cms/sections/:id/edit|admin/api/cms/sections#edit|
|GET|/admin/api/cms/sections/:id|admin/api/cms/sections#show|
|PATCH|/admin/api/cms/sections/:id|admin/api/cms/sections#update|
|PUT|/admin/api/cms/sections/:id|admin/api/cms/sections#update|
|DELETE|/admin/api/cms/sections/:id|admin/api/cms/sections#destroy|
|GET|/admin/api/cms/files|admin/api/cms/files#index|
|POST|/admin/api/cms/files|admin/api/cms/files#create|
|GET|/admin/api/cms/files/new|admin/api/cms/files#new|
|GET|/admin/api/cms/files/:id/edit|admin/api/cms/files#edit|
|GET|/admin/api/cms/files/:id|admin/api/cms/files#show|
|PATCH|/admin/api/cms/files/:id|admin/api/cms/files#update|
|PUT|/admin/api/cms/files/:id|admin/api/cms/files#update|
|DELETE|/admin/api/cms/files/:id|admin/api/cms/files#destroy|
|PUT|/admin/api/cms/templates/:id/publish|admin/api/cms/templates#publish|
|GET|/admin/api/cms/templates|admin/api/cms/templates#index|
|POST|/admin/api/cms/templates|admin/api/cms/templates#create|
|GET|/admin/api/cms/templates/:id|admin/api/cms/templates#show|
|PATCH|/admin/api/cms/templates/:id|admin/api/cms/templates#update|
|PUT|/admin/api/cms/templates/:id|admin/api/cms/templates#update|
|DELETE|/admin/api/cms/templates/:id|admin/api/cms/templates#destroy|

**Special notes for your reviewer**:

### CODE REVIEW

**Review by jlledom — COMMENTED** [review]
There are some endpoints in your list that don't exist AFAIK:

| Verb | Path | Action |
|---|---|---|
|GET| 	/admin/api/cms/sections/new| 	admin/api/cms/sections#new|
|GET| 	/admin/api/cms/sections/:id/edit| 	admin/api/cms/sections#edit|
|GET| 	/admin/api/cms/files/new| 	admin/api/cms/files#new|
|GET| 	/admin/api/cms/files/:id/edit| 	admin/api/cms/files#edit|

Are those routes valid and working? because there's no action in the controllers for them.

Also, I didn't even know this one existed:

| Verb | Path | Action |
|---|---|---|
|GET| 	/admin/api/cms/sections/:section_id/files| 	admin/api/cms/files#index|

I think we should remove it, if we provide `sections/:section_id/files` we should also provide `sections/:section_id/sections` and `sections/:section_id/templates`. And we have an issue ([THREESCALE-9191](https://issues.redhat.com/browse/THREESCALE-9191)) to filter results by any parameter, so the client will be able to retrieve the same results by filtering by `section_id`.

File: app/controllers/admin/api/cms/base_controller.rb [hunk 1]
```diff
@@ -1,5 +1,7 @@
-class Admin::Api::CMS::BaseController < Admin::Api::BaseController
+# frozen_string_literal: true
 
+class Admin::Api::CMS::BaseController < Admin::Api::BaseController
+  before_action :ensure_json_request
```
[the remark this answers is not in the document]
Comment (thalesmiguel) [reply app/controllers/admin/api/cms/base_controller.rb:4]: We could add `constraints: { format: 'json' }` to the [CMS routes](https://github.com/3scale/porta/pull/3255/files#diff-959bc9abc46a55332bb64d5155a79323afa75a50ec1a2137ddd22d926f62c6c5R553) instead but that would return a 404 for requests that specifically ask for a different format (`.xml`, `.html`, ...).
From the API perspective, responding with a 406 or even a 422 would be more meaningful.
I guess the question here is how we want to communicate users that the CMS api only responds to `json`.

File: app/models/cms/builtin.rb [hunk 2]
```diff
@@ -110,52 +110,13 @@ class StaticPage < CMS::Builtin
...
-          xml.updated_at updated_at.xmlschema
-        end
-
-        x.system_name system_name
-        x.liquid_enabled liquid_enabled
-        x.layout layout_name
-      end
-
-      xml.to_xml
-    end
```
[the remark this answers is not in the document]
Comment (thalesmiguel) [reply app/models/cms/builtin.rb:130]: As sure as the tests let us be 😅.
Everything is green and I didn't have any issues in my manual tests.

File: test/integration/cms/api/templates_test.rb [hunk 3]
```diff
@@ -18,17 +17,16 @@ def setup
...
+        templates = response.parsed_body['templates']
 
-        assert_equal 4, doc.xpath('/templates/*').size
-        assert_equal 1, doc.xpath('/templates/partial').size
-        assert_equal 0, doc.xpath('/templates/partial/draft').size
-        assert_equal 1, doc.xpath('/templates/page').size
-        assert_equal 1, doc.xpath('/templates/builtin_page').size
-        assert_equal 1, doc.xpath('/templates/layout').size
+        assert_equal 4, templates.size
+        assert_equal 1, templates.find { |templ| templ['partial'] }.values.size
```
[the remark this answers is not in the document]
Comment (thalesmiguel) [reply test/integration/cms/api/templates_test.rb:26]: Added in faf6db3.

**Comment by thalesmiguel** [comment]
> There are some endpoints in your list that don't exist AFAIK:

Yep... Those are "new" and "edit" endpoints that come from the "resources" Rails routes helper. Those don't really make sense for an API. 

> Also, I didn't even know this one existed:
> I think we should remove it, if we provide `sections/:section_id/files` we should also provide `sections/:section_id/sections` and `sections/:section_id/templates`. And we have an issue ([THREESCALE-9191](https://issues.redhat.com/browse/THREESCALE-9191)) to filter results by any parameter, so the client will be able to retrieve the same results by filtering by `section_id`.

Agreed!

Since those changes go out of the scope of the PR, I opened https://github.com/3scale/porta/pull/3266 to remove the endpoints you mentioned @jlledom :v:

File: app/controllers/admin/api/cms/base_controller.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (jlledom) [reply app/controllers/admin/api/cms/base_controller.rb:4]: OK, let's use the callback then.

**Review by jlledom — APPROVED** [review]

**Pull request merged**